### PR TITLE
Add support for DW_TAG_class_type 

### PR DESCRIPTION
**Pull request by amlannayak**
drgn currently is not equipped to handle class types making it difficult to use with C++ binaries. Here we're adding the initial support necessary to enable class type parsing. Since class and struct types are nearly identical in their DWARF representation, we are simply extending the logic already defined for structs to the `DW_TAG_class_type`. 

Note that since some methods use the `is_struct` boolean parameter to differentiate between `structs` and `unions`, we've added an additional bool to account for the `class` case. But this could probably be better captured with an `enum`.

### CODE REVIEW

File: libdrgn/cityhash.h [hunk 1]
```diff
@@ -70,7 +70,8 @@ static uint32_t cityhash32_len_13to24(const uint8_t *s, size_t len) {
 static uint32_t cityhash32_len_0to4(const uint8_t *s, size_t len) {
 	uint32_t b = 0;
 	uint32_t c = 9;
-	for (size_t i = 0; i < len; i++) {
+  size_t i = 0;
```
Comment (amlannayak) [inline libdrgn/cityhash.h]: This was due to a compilation error. Can revert it.

File: libdrgn/language_c.c [hunk 2]
```diff
@@ -19,6 +19,8 @@
 #include "type.h"
 #include "type_index.h"
 
+#define DBL_DECIMAL_DIG 17
```
Comment (amlannayak) [inline libdrgn/language_c.c]: Also due to a compilation error. gcc wasn't able to find the definition of this macro even though it should technically exist in `float.h`

File: libdrgn/dwarf_info_cache.c [hunk 3]
```diff
@@ -543,7 +544,7 @@ static struct drgn_error *parse_member(struct drgn_dwarf_info_cache *dicache,
 
 static struct drgn_error *
 drgn_compound_type_from_dwarf(struct drgn_dwarf_info_cache *dicache,
-			      Dwarf_Die *die, bool is_struct,
+			      Dwarf_Die *die, bool is_struct, bool is_class,
```
[the remark this answers is not in the document]
Comment (amlannayak) [reply libdrgn/dwarf_info_cache.c]: Agreed. Switching to the enum to make it cleaner.

File: libdrgn/drgn.h [hunk 4]
```diff
@@ -287,6 +287,8 @@ enum drgn_type_kind {
 	DRGN_TYPE_FLOAT,
 	/** Complex type. */
 	DRGN_TYPE_COMPLEX,
+  /** Class type. */
+  DRGN_TYPE_CLASS,
```
[the remark this answers is not in the document]
Comment (amlannayak) [reply libdrgn/drgn.h]: Fixed.

**Comment by amlannayak** [comment]
Added tests and addressed comments.

**Comment by osandov** [comment]
Thanks, this looks good overall. I still had to fix up some whitespace, but other than that I'm gonna do some testing and then push it out.

**Comment by osandov** [comment]
I pushed a few changes to your branch:
* I rebased on the master branch and flattened all of the merge commits.
* I documented `class_type` in the API reference, which fixed a build error that `drgn_class_type_DOC` wasn't defined.
* I fixed some typos in the test cases and expanded them a bit.
* I cleaned up the whitespace and other miscellaneous things.
* Most importantly, I removed the `class` keyword parsing from `language_c.c`. `class` isn't a keyword in C, so we still need to treat it as an identifier. In fact, the Linux kernel has a `struct class`. We'll need to have C++-specific language callbacks for that functionality.

**Comment by osandov** [comment]
The C++-specific stuff should be a new pull request. I think what we have now is good enough to merge, but please double-check my version when you get the chance.

**Comment by amlannayak** [comment]
Thanks for the fixes! I pulled this locally and tested. Everything seems to be good. 

I wasn't aware of the `struct class` structure in the kernel. So, what would be proper approach to handle `class` types in a C++-specific context?

**Comment by osandov** [comment]
We'll need a separate `struct drgn_language drgn_language_cpp` with a separate `cpp_find_type` that knows about the `class` token and can find either `class Foo` or `Foo` (although we'll want to share as much code as possible between that and `drgn_language_c`, since they'll be mostly the same). Then we'll need to assign the language of a program appropriately when we get debug info for it.

**Comment by osandov** [comment]
I merged the class type support, thanks!

**Pull request merged**